src/library/export/engineprimeexportjob.cpp

```diff
@@ -0,0 +1,532 @@
+    externalTrack.set_loops({});
+
+    // Write waveform.
+    // Note that writing a single waveform will automatically calculate an
+    // overview waveform too.
+    if (pWaveform) {
+        int64_t samplesPerEntry = externalTrack.required_waveform_samples_per_entry();
+        int64_t externalWaveformSize = (sampleCount + samplesPerEntry - 1) / samplesPerEntry;
+        std::vector<djinterop::waveform_entry> externalWaveform;
+        externalWaveform.reserve(externalWaveformSize);
+        for (int64_t i = 0; i < externalWaveformSize; ++i) {
+            auto j = pWaveform->getDataSize() * i / externalWaveformSize;
+            externalWaveform.push_back({{pWaveform->getLow(j), 127},
+                    {pWaveform->getMid(j), 127},
+                    {pWaveform->getHigh(j), 127}});
+        }
+        externalTrack.set_waveform(std::move(externalWaveform));
+    } else {
+        qInfo() << "No waveform data found for track" << pTrack->getId()
+                << "(" << pTrack->getFileInfo().fileName() << ")";
+    }
+}
+
+void exportTrack(
+        const EnginePrimeExportRequest& request,
+        djinterop::database& db,
+        QHash<TrackId, int64_t>& mixxxToEnginePrimeTrackIdMap,
+        const TrackPointer pTrack,
+        std::unique_ptr<Waveform> pWaveform) {
+    // Only export supported file types.
+    if (!kSupportedFileTypes.contains(pTrack->getType())) {
+        qInfo() << "Skipping file" << pTrack->getFileInfo().fileName()
+                << "(id" << pTrack->getId() << ") as its file type"
+                << pTrack->getType() << "is not supported";
+        return;
+    }
+
+    // Copy the file, if required.
+    auto musicFileRelativePath = exportFile(request, pTrack);
+
+    // Export meta-data.
+    exportMetadata(db,
+            mixxxToEnginePrimeTrackIdMap,
+            pTrack,
+            std::move(pWaveform),
+            musicFileRelativePath);
+}
+
+void exportCrate(
+        djinterop::crate& extRootCrate,
+        QHash<TrackId, int64_t>& mixxxToEnginePrimeTrackIdMap,
+        const Crate& crate,
+        const QList<TrackId>& trackIds) {
+    // Create a new crate as a sub-crate of the top-level Mixxx crate.
+    auto extCrate = extRootCrate.create_sub_crate(crate.getName().toStdString());
+
+    // Loop through all track ids in this crate and add.
+    for (auto trackId : trackIds) {
+        auto extTrackId = mixxxToEnginePrimeTrackIdMap[trackId];
+        extCrate.add_track(extTrackId);
+    }
+}
+
+} // namespace
+
+EnginePrimeExportJob::EnginePrimeExportJob(
+        QObject* parent,
+        TrackCollectionManager* pTrackCollectionManager,
+        EnginePrimeExportRequest request)
+        : QThread(parent),
+          m_pTrackCollectionManager(pTrackCollectionManager),
+          m_request{std::move(request)} {
+    // Must be collocated with the TrackCollectionManager.
+    DEBUG_ASSERT(m_pTrackCollectionManager != nullptr);
+    moveToThread(m_pTrackCollectionManager->thread());
```

Comment:
moveToThread() will fail if `parent`'s thread and the current thread are different. All member of parent/child hierarchy must live in the same thread.

Reply:
To clarify, are you suggesting that it is a good idea to assert the thread affinity in the ctor before the call to `moveToThread()`?

```cpp
DEBUG_ASSERT(m_pTrackCollectionManager != nullptr);
DEBUG_ASSERT_QOBJECT_THREAD_AFFINITY(parent);
moveToThread(m_pTrackCollectionManager->thread());
```

Reply:
No.

> The object cannot be moved if it has a parent.

https://doc.qt.io/qt-5/qobject.html#moveToThread

Reply:
DEBUG_ASSERT_QOBJECT_THREAD_AFFINITY(m_pTrackCollectionManager)

The object must be constructed within the same thread context as m_pTrackCollectionManager, because moveToThread() won't work for parented objects.

Reply:
Alternatively omit the parent, whatever is more suitable.

Reply:
Since it is known to be constructed from the main thread, and that `m_pTrackCollectionManager` is also on that thread, I've gone for adding `DEBUG_ASSERT_QOBJECT_THREAD_AFFINITY(m_pTrackCollectionManager)` to the constructor.

Reply:
Slight adjustment: I've instead put in a switch so that the ctor will either move itself to the right thread (if constructed without a parent), or assert that it is already in the right thread.